TorchSharp/FileOperations.tt

```diff
@@ -62,10 +62,48 @@ foreach (var type in new [] {
         ///   Write <#=accreal#>s to the file from the given storage.
         /// </summary>
         /// <param name="storage">A storage object fetch data from.</param>
-        /// <returns>The number of bytes written.</returns>
-        public long Write<#=tname#>(<#=tname#>Tensor.<#=tname#>Storage storage) { return THFile_write<#=tname#>(this.handle, storage.handle); }
+        /// <returns>The number of <#=accreal#>s written.</returns>
+        public long Write<#=tname#>s(<#=tname#>Tensor.<#=tname#>Storage storage) { return THFile_write<#=tname#>(this.handle, storage.handle); }
+
+        [DllImport("caffe2")] 
+        extern static long THFile_read<#=tname#>Raw(HType self, IntPtr data, long n);
+        
+        /// <summary>
+        ///   Read <#=accreal#>s from the file into the given <#=accreal#> array.
+        /// </summary>
+        /// <param name="data">An array to place the data in after reading it from the file.</param>
+        /// <param name="n">The maximum number of <#=accreal#>s to read.</param>
+        /// <returns>The number of <#=accreal#>s read.</returns>
+        public long Read<#=tname#>s(<#=accreal#>[] data, int n)
+        {
+            if (n > data.Length)
+                throw new ArgumentOutOfRangeException("n cannot be greater than data.Length");
+            var dest = Marshal.AllocHGlobal(n*sizeof(<#=accreal#>));
+            var readItems = THFile_read<#=tname#>Raw(this.handle, dest, n);            
+            Marshal.Copy(dest, data, 0, (int)readItems);
```

Comment:
Why do we do a copy here? For the unsafe thing? I prefer to do unsafe than copying.

Reply:
The input is a managed array. I guess I don't know how to convert it to a double*

Reply:
You can use somethign like:

```
unsafe {
fixed (<#=accreal#> *ptr = data) {
    // Use the ptr here.
}
}
```

Reply:
Okay, I've fixed this.